### PR TITLE
Continue with forward proxy if ReverseOnly is not true and no mapping available

### PR DESCRIPTION
**Pull request by sleemburg**
Hi,

As indicated by Michael Adam, a new pull request with rewritten code.

Kind regards,
Stephan

### CODE REVIEW

File: src/reqs.c [hunk 1]
```diff
@@ -369,12 +369,15 @@ static struct request_s *process_request (struct conn_s *connptr,
 
                 reverse_url = reverse_rewrite_url (connptr, hashofheaders, url);
 
-                if (!reverse_url) {
+                if (reverse_url != NULL) {
+                        safefree (url);
+                        url = reverse_url;
+		} else if (config.reverseonly) {
```
Comment (obnoxxx) [inline src/reqs.c]: There is a whitespace error here: TP currently uses spaces, not tabs.

File: src/reqs.c [hunk 2]
```diff
@@ -369,12 +369,15 @@ static struct request_s *process_request (struct conn_s *connptr,
...
+                        indicate_http_error (connptr, 400, "Bad Request",
+                                     "detail", "No mapping found for requested url",
```
Comment (obnoxxx) [inline src/reqs.c]: Formatting: please indent the arguments to match the opening bracket, and make sure the line length does not exceed 80 characters, if possible. i.e.

```
    func (connptr, ...
          "detail", "No ...."
          ...);
```

**Comment by obnoxxx** [comment]
The patch looks good in general, but see the formatting comments in the patch.

also copying the comment from #32:



The diff looks good now. The two patches named "allow non-reverse mappings if reverseonly is not enabled" should be squashed. But I think for correctness the patch should modify reverse_rewrite_url() in src/reverse-proxy.c:160ff to remove the indicate_http_error call there and also the check for config.reverseonly. so that the hunk reads roughly like this:

```
   if (rewrite_url == NULL) {
        return NULL;
   }
```
because you put all the 'reverse_only' logic into the caller. am I correct?

**Comment by sleemburg** [comment]
I made some changes, is it ok now?

File: src/reverse-proxy.c [hunk 3]
```diff
@@ -166,7 +166,8 @@ char *reverse_rewrite_url (struct conn_s *connptr, hashmap_t hashofheaders,
                 return NULL;
         }
 
-        log_message (LOG_CONN, "Rewriting URL: %s -> %s", url, rewrite_url);
+        log_message (LOG_CONN, "Rewriting URL: %s -> %s", url,
```
Comment (obnoxxx) [inline src/reverse-proxy.c]: There is a trailing space at the end of the line, after ` url, `. That should be removed.

File: src/reverse-proxy.c [hunk 4]
```diff
@@ -166,7 +166,8 @@ char *reverse_rewrite_url (struct conn_s *connptr, hashmap_t hashofheaders,
                 return NULL;
         }
 
-        log_message (LOG_CONN, "Rewriting URL: %s -> %s", url, rewrite_url);
+        log_message (LOG_CONN, "Rewriting URL: %s -> %s", url, 
+			rewrite_url ? rewrite_url : "No mapping found");
```
Comment (obnoxxx) [inline src/reverse-proxy.c]: The indentation is done with tabs here. Should be spaces to be consistent with the rest of the code.
And should be aligned with the opening bracket.

**Comment by obnoxxx** [comment]
There are still formatting issues, and see the comment for the proposed change to reverse_rewrite_url() to remove the call to `indicate_http_error` from `reverse_rewrite_url`. I could do that for you, but want to give you the opportunity to do the changes yourself. This iterative review<->amend-patch cycle is quite common.

**Comment by sleemburg** [comment]
I changed it, but forgot to do a push origin after the commit.. Still used to subversion..

I removed the indicate_http_error and the formatting with the tabs and null dereference is undone, as the null dereference is not possible anymore with the current flow.

Thank you for your help. I prefer to do the changes myself and learn from this review/amend-patch cycle.

File: src/reqs.c [hunk 5]
```diff
@@ -372,10 +372,13 @@ static struct request_s *process_request (struct conn_s *connptr,
                 if (reverse_url != NULL) {
                         safefree (url);
                         url = reverse_url;
-		} else if (config.reverseonly) {
+                } else if (config.reverseonly) {
+                        log_message (LOG_ERR,
```
Comment (obnoxxx) [inline src/reqs.c]: There is 1 trailing space at the end. Please remove.

File: src/reqs.c [hunk 6]
```diff
@@ -372,10 +372,13 @@ static struct request_s *process_request (struct conn_s *connptr,
                 if (reverse_url != NULL) {
                         safefree (url);
                         url = reverse_url;
-		} else if (config.reverseonly) {
+                } else if (config.reverseonly) {
+                        log_message (LOG_ERR, 
+                                    "Bad request, no mapping for '%s' found",
+                                    url);
```
Comment (obnoxxx) [inline src/reqs.c]: These need 1 space more of indentation: should be aligned not below the bracket but below the item in the bracket.

**Comment by obnoxxx** [comment]
Thanks for updating! It's almost perfect now. Just two more whitespace errors.
The logic looks very clean and correct now.
Also good catch that the null-deref check is now not needed any more.

If you fix the whitespace errors, I'll be happy and merge.

File: src/reqs.c [hunk 7]
```diff
@@ -373,7 +373,7 @@ static struct request_s *process_request (struct conn_s *connptr,
                         safefree (url);
                         url = reverse_url;
                 } else if (config.reverseonly) {
-                        log_message (LOG_ERR, 
+                        log_message (LOG_ERR,
```
Comment (obnoxxx) [inline src/reqs.c]: Great!

Now only indent the two following lines one more space and we're done. ;-)

**Comment by sleemburg** [comment]
I fixed the whitespaces.

File: src/reqs.c [hunk 8]
```diff
@@ -374,8 +374,8 @@ static struct request_s *process_request (struct conn_s *connptr,
                         url = reverse_url;
                 } else if (config.reverseonly) {
                         log_message (LOG_ERR,
-                                    "Bad request, no mapping for '%s' found",
-                                    url);
+                                     "Bad request, no mapping for '%s' found",
+                                     url);
```
Comment (obnoxxx) [inline src/reqs.c]: perfect.

**Comment by obnoxxx** [comment]
Merged.
Thanks a lot for your contribution!
Tinyproxy has become a bit better! :-)

**Comment by sleemburg** [comment]
Thanks! Also for your help.

**Pull request merged**